sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelineEventSender.scala

```diff
@@ -0,0 +1,194 @@
+    sessionHolder: SessionHolder)
+    extends Logging {
+
+  // Thread-safe queue for buffering events
+  private val eventQueue: BlockingQueue[PipelineEventMessage] = new LinkedBlockingQueue()
+
+  // ExecutorService for background event processing
+  private val executor: ExecutorService = {
+    val threadFactory = ThreadUtils
+      .namedThreadFactory(s"pipeline-event-sender-${sessionHolder.sessionId}")
+    Executors.newSingleThreadExecutor(threadFactory)
+  }
+
+  // state tracking
+  private val isShutdown = new AtomicBoolean(false)
+  private val isStarted = new AtomicBoolean(false)
+
+  // Sealed trait for different message types
+  private sealed trait PipelineEventMessage
+  private case class EventMessage(event: PipelineEvent) extends PipelineEventMessage
+  private case object ShutdownMessage extends PipelineEventMessage
+
+  /**
+   * Start the background executor for sending events, only if not already started.
+   */
+  def start(): Unit = {
+    if (isStarted.compareAndSet(false, true)) {
+      executor.submit(new Runnable {
+        override def run(): Unit = {
+          try {
+            processEvents()
+          } catch {
+            case NonFatal(e) =>
+              logError(s"Pipeline event sender failed for session ${sessionHolder.sessionId}", e)
+          }
+        }
+      })
+      logInfo(s"Started pipeline event sender for session ${sessionHolder.sessionId}")
+    }
+  }
+
+  /**
+   * Send an event asynchronously by adding it to the queue
+   */
+  def sendEvent(event: PipelineEvent): Unit = {
+    if (!isShutdown.get()) {
+      val offered = eventQueue.offer(EventMessage(event))
+      if (!offered) {
+        logWarning(s"Failed to queue pipeline event for session ${sessionHolder.sessionId}")
+      }
+    }
+  }
+
+  /**
+   * Shutdown the event sender, stop taking new events and wait for processing to complete.
+   */
+  def shutdown(): Unit = {
+    if (isShutdown.compareAndSet(false, true)) {
+      // Signal shutdown to the processing loop
+      eventQueue.offer(ShutdownMessage)
+      // Request a shutdown of the executor which waits for processEvents to complete
+      executor.shutdown()
+      // Blocks until all tasks have completed execution after a shutdown request,
+      // disregard the timeout since we want all events to be processed
+      if (!executor.awaitTermination(Long.MaxValue, java.util.concurrent.TimeUnit.MILLISECONDS)) {
+        logError(s"Pipeline event sender for session ${sessionHolder.sessionId}" +
+          s"failed to terminate")
+        executor.shutdownNow()
+      }
+    }
+  }
+
+  /**
+   * Main event processing loop that runs in the background executor
+   */
+  private def processEvents(): Unit = {
+    while (true) {
+      try {
+        val message = eventQueue.take()
+
+        message match {
+          case EventMessage(event) =>
+            sendEventToClient(event)
+          case ShutdownMessage =>
```

Comment:
Do we really need this ShutdownMessage? Can the shutdown call not just interrupt this thread?